daffodil-macro-lib/src/main/scala/org/apache/daffodil/processors/parsers/PointOfUncertaintyMacros.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.processors.parsers
+
+import scala.reflect.macros.blackbox.Context
+
+
+object PointOfUncertaintyMacros {
+
+  /**
+   * For creating a point of uncertainty and ensuring that it is cleaned up
+   * appropriately. This allows the use of standard scala-style syntax to
+   * create and use points of uncertainties, while avoiding allocating
+   * functions. Thus, this allows something like this:
+   *
+   *   pstate.withPointOfUncertainty { pou =>
+   *     // code that implements parsing, and may potentially reset to,
+   *     // discard, or resolve this pou variable
+   *   }
+   *
+   * Upon completion of the block, if the new point of uncertainty has not been
+   * discarded, reset to, or resovled, this pou will then be discarded,
+   * ensuring that Marks are always cleaned up appropriately.
+   */
+  def withPointOfUncertainty[A, B](c: Context)(pouID: c.Expr[String], context: c.Tree)(func: c.Expr[A => B]) = {
+
+    import c.universe._
+    
+    val state = TermName(c.freshName("state"))
+    val id = TermName(c.freshName("id"))
+    val pou = TermName(c.freshName("pou"))
+    val ctx = TermName(c.freshName("context"))
+
+    func.tree match {
+
+      // The func tree is something like (param => body), where param is the
+      // name the caller wants to name the point of uncertainty mark, and body
+      // is the code that uses this point of uncertainty. We transform this
+      // code to create a new point of uncertainty, evaluate the body, and then
+      // use a finally block to ensure we discard the PoU if it has not been
+      // resolved.
+      case q"""($param: $_) => $body""" => {
+
+        // The "body" uses the "param" variable name, and scala doesn't like it
+        // if we just create a new variable with this name, with a very unclear
+        // error message. Instead, we need to create a "fresh" variable name to
+        // hold he new point of uncertainty, and then transform the body so
+        // that all instances of "param" are replaced with our new fresh name.
```

Comment:
This seems pretty fragile. E.g., reflection to a name will clearly break this. 
I would add a big scaladoc warning so that people keep the scope over which this transform is applied small.

Reply:
Agreed. Will do.